File: .github/scripts/fix-code.js
const { Octokit } = require('@octokit/rest');
const axios = require('axios');
const fs = require('fs');
const { execSync } = require('child_process');
const path = require('path');

// Initialize GitHub client
const octokit = new Octokit({
  auth: process.env.GITHUB_TOKEN
});

// Parse environment variables
const [owner, repo] = process.env.REPO.split('/');
const commitSha = process.env.COMMIT_SHA;
const lintErrors = JSON.parse(process.env.LINT_ERRORS || '{}');
const testErrors = JSON.parse(process.env.TEST_ERRORS || '{}');
const affectedFiles = JSON.parse(process.env.AFFECTED_FILES || '[]');
const maxAttempts = parseInt(process.env.MAX_ATTEMPTS || '3');

// Keep track of original errors for reporting
const originalErrors = {
  lint: lintErrors,
  test: testErrors
};

// Main function to fix code issues
async function fixCode() {
  try {
    console.log('Starting auto-fix process...');
    console.log(`Affected files: ${JSON.stringify(affectedFiles)}`);
    console.log(`Lint errors: ${JSON.stringify(lintErrors)}`);
    console.log(`Test errors: ${JSON.stringify(testErrors)}`);
    
    let success = false;
    let changesLog = [];
    let verificationResults = '';
    
    // Combine all files that need fixing
    const filesToFix = new Set([
      ...Object.keys(lintErrors),
      ...Object.keys(testErrors)
    ].filter(file => file !== '_general'));
    
    // If we have general test errors with no specific files, try to identify files from affected files
    if (testErrors['_general'] && affectedFiles.length > 0) {
      affectedFiles.forEach(file => filesToFix.add(file));
    }
    
    console.log(`Files to fix: ${JSON.stringify([...filesToFix])}`);
    
    for (const filePath of filesToFix) {
      if (!fs.existsSync(filePath)) {
        console.log(`File ${filePath} does not exist, skipping.`);
        continue;
      }
      
      const fileContent = fs.readFileSync(filePath, 'utf8');
      const fileExtension = path.extname(filePath);
      
      // Determine file type based on extension
      let fileType;
      if (['.js', '.jsx', '.ts', '.tsx'].includes(fileExtension)) {
        fileType = 'JavaScript/TypeScript';
      } else if (['.py'].includes(fileExtension)) {
        fileType = 'Python';
      } else if (['.java'].includes(fileExtension)) {
        fileType = 'Java';
      } else if (['.rb'].includes(fileExtension)) {
        fileType = 'Ruby';
      } else if (['.go'].includes(fileExtension)) {
        fileType = 'Go';
      } else if (['.php'].includes(fileExtension)) {
        fileType = 'PHP';
      } else if (['.c', '.cpp', '.h', '.hpp'].includes(fileExtension)) {
        fileType = 'C/C++';
      } else if (['.cs'].includes(fileExtension)) {
        fileType = 'C#';
      } else if (['.html', '.htm'].includes(fileExtension)) {
        fileType = 'HTML';
      } else if (['.css', '.scss', '.sass', '.less'].includes(fileExtension)) {
        fileType = 'CSS';
      } else {
        fileType = 'Unknown';
      }
      
      const fileErrors = [];
      
      // Collect lint errors for this file
      if (lintErrors[filePath]) {
        fileErrors.push(...lintErrors[filePath].map(err => 
          `Lint error at line ${err.line}${err.column ? `, column ${err.column}` : ''}: ${err.message}`
        ));
      }
      
      // Collect test errors for this file
      if (testErrors[filePath]) {
        fileErrors.push(...testErrors[filePath].map(err => 
          `Test error: ${err.message} - ${err.details}`
        ));
      }
      
      // If there are general test errors, include them for all files we're fixing
      if (testErrors['_general']) {
        fileErrors.push(...testErrors['_general'].map(err => 
          `General test error that might relate to this file: ${err.message}`
        ));
      }
      
      if (fileErrors.length === 0) {
        console.log(`No specific errors found for ${filePath}, skipping.`);
        continue;
      }
      
      console.log(`Attempting to fix ${filePath}...`);
      
      let fixed = false;
      let attemptsMade = 0;
      let currentFileContent = fileContent;
      
      while (!fixed && attemptsMade < maxAttempts) {
        attemptsMade++;
        console.log(`Attempt ${attemptsMade}/${maxAttempts} for ${filePath}`);
        
        try {
          // Get fix suggestions from OpenAI
          const fixedContent = await getFixFromOpenAI(currentFileContent, fileErrors, fileType, filePath, attemptsMade);
          
          if (!fixedContent || fixedContent === currentFileContent) {
            console.log('No changes suggested or OpenAI returned the same content');
            if (attemptsMade === maxAttempts) {
              console.log(`Maximum attempts reached for ${filePath} without success`);
            }
            continue;
          }
          
          // Save the fixed content
          fs.writeFileSync(filePath, fixedContent);
          console.log(`Updated ${filePath} with suggested fixes`);
          
          // Commit the changes
          try {
            execSync(`git add "${filePath}"`);
            execSync(`git commit -m "fix: Auto-fix attempt ${attemptsMade} for ${filePath}"`);
            console.log(`Committed changes for ${filePath}`);
            
            // Log the changes
            const diff = execSync(`git diff HEAD~1 HEAD -- "${filePath}"`).toString();
            changesLog.push(`## Changes to ${filePath} (Attempt ${attemptsMade})\n\`\`\`diff\n${diff}\n\`\`\``);
            
            // Verify the changes by running relevant checks
            const verificationResult = await verifyFix(filePath, fileType);
            console.log(`Verification result: ${JSON.stringify(verificationResult)}`);
            
            if (verificationResult.success) {
              console.log(`Fix for ${filePath} was successful!`);
              fixed = true;
              success = true;
              verificationResults += `\n### ${filePath} - ✅ Fixed successfully\n${verificationResult.details}\n`;
            } else {
              console.log(`Fix for ${filePath} did not resolve all issues, trying again...`);
              // Update current content for next attempt
              currentFileContent = fixedContent;
              // Update file errors based on verification result
              if (verificationResult.errors && verificationResult.errors.length > 0) {
                fileErrors.length = 0; // Clear previous errors
                fileErrors.push(...verificationResult.errors);
              }
              verificationResults += `\n### ${filePath} - ❌ Attempt ${attemptsMade} failed\n${verificationResult.details}\n`;
            }
          } catch (commitError) {
            console.error(`Error committing changes: ${commitError}`);
            verificationResults += `\n### ${filePath} - ❌ Error committing changes\n${commitError}\n`;
          }
        } catch (error) {
          console.error(`Error fixing ${filePath}: ${error}`);
          verificationResults += `\n### ${filePath} - ❌ Error in fix attempt ${attemptsMade}\n${error}\n`;
        }
      }
      
      if (!fixed) {
        console.log(`Could not fix ${filePath} after ${maxAttempts} attempts`);
      }
    }
    
    // Set the outputs
    console.log(`::set-output name=success::${success}`);
    console.log(`::set-output name=original-errors::${JSON.stringify(originalErrors)}`);
    console.log(`::set-output name=changes-made::${changesLog.join('\n\n')}`);
    console.log(`::set-output name=verification-results::${verificationResults}`);
    
    if (success) {
      console.log('Successfully fixed at least one file!');
      process.exit(0);
    } else {
      console.log('Failed to fix any files.');
      process.exit(1);
    }
  } catch (error) {
    console.error('Error in fixCode function:', error);
    process.exit(1);
  }
}

// Function to get fix suggestions from OpenAI
async function getFixFromOpenAI(fileContent, errors, fileType, filePath, attempt) {
  try {
    console.log(`Requesting OpenAI fix for ${filePath} (Attempt ${attempt})`);
    
    const errorText = errors.join('\n');
    
    // Construct a prompt for the OpenAI API
    const prompt = `You are an expert ${fileType} developer. I need your help fixing errors in a file.

File Path: ${filePath}
File Type: ${fileType}
Attempt: ${attempt}/${maxAttempts}

The file has the following errors:
${errorText}

Here is the current content of the file:
\`\`\`
${fileContent}
\`\`\`

Please provide ONLY the fixed version of the file with no explanation. Your response should be the complete file content that resolves the errors.`;

    const response = await axios.post(
      'https://api.openai.com/v1/chat/completions',
      {
        model: 'gpt-3.5-turbo',
        messages: [{
          role: 'user',
          content: prompt
        }],
        temperature: 0.2,
        max_tokens: 4096
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`
        }
      }
    );
    
    if (response.data && response.data.choices && response.data.choices.length > 0) {
      let fixedContent = response.data.choices[0].message.content;
      
      // Strip out markdown code blocks if present
      if (fixedContent.includes('```')) {
        const codeBlockMatch = fixedContent.match(/```(?:\w+)?\n([\s\S]+?)```/);
        if (codeBlockMatch && codeBlockMatch[1]) {
          fixedContent = codeBlockMatch[1];
        } else {
          // Try another pattern
          const lines = fixedContent.split('\n');
          const startIdx = lines.findIndex(line => line.startsWith('```'));
          const endIdx = lines.slice(startIdx + 1).findIndex(line => line.startsWith('```')) + startIdx + 1;
          
          if (startIdx !== -1 && endIdx !== -1) {
            fixedContent = lines.slice(startIdx + 1, endIdx).join('\n');
          }
        }
      }
      
      return fixedContent;
    } else {
      console.error('Unexpected response format from OpenAI');
      return null;
    }
  } catch (error) {
    console.error('Error getting fix from OpenAI:', error.response?.data || error.message);
    return null;
  }
}

// Function to verify if the fix resolves the issues
async function verifyFix(filePath, fileType) {
  try {
    console.log(`Verifying fix for ${filePath}`);
    
    // Determine the appropriate verification command based on file type
    let command = '';
    let useGenericLint = false;
    
    if (['.js', '.jsx'].includes(path.extname(filePath))) {
      if (fs.existsSync('package.json')) {
        const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
        if (packageJson.scripts) {
          if (packageJson.scripts.lint) {
            command = 'npm run lint';
          } else if (packageJson.scripts.eslint) {
            command = 'npm run eslint';
          } else {
            useGenericLint = true;
          }
        }
      }
      
      if (useGenericLint || !command) {
        if (fs.existsSync('.eslintrc') || fs.existsSync('.eslintrc.js') || fs.existsSync('.eslintrc.json')) {
          command = `npx eslint "${filePath}"`;
        }
      }
    } else if (['.ts', '.tsx'].includes(path.extname(filePath))) {
      if (fs.existsSync('package.json')) {
        const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
        if (packageJson.scripts && packageJson.scripts.lint) {
          command = 'npm run lint';
        } else {
          useGenericLint = true;
        }
      }
      
      if (useGenericLint || !command) {
        if (fs.existsSync('tsconfig.json')) {
          command = `npx tsc --noEmit`;
        }
      }
    } else if (path.extname(filePath) === '.py') {
      if (fs.existsSync('requirements.txt') || fs.existsSync('pyproject.toml')) {
        command = `python -m pylint "${filePath}"`;
      }
    }
    
    // If we couldn't determine a specific linting command, try running tests
    if (!command) {
      if (fs.existsSync('package.json')) {
        const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
        if (packageJson.scripts && packageJson.scripts.test) {
          command = 'npm test';
        }
      } else if (fs.existsSync('pytest.ini') || fs.existsSync('conftest.py')) {
        command = 'python -m pytest';
      } else if (fs.existsSync('build.gradle') || fs.existsSync('build.gradle.kts')) {
        command = './gradlew test';
      }
    }
    
    if (!command) {
      return {
        success: false,
        details: 'Could not determine appropriate verification command for this file type.',
        errors: []
      };
    }
    
    console.log(`Running verification command: ${command}`);
    
    try {
      const output = execSync(command, { stdio: 'pipe' }).toString();
      console.log(`Verification succeeded with output: ${output}`);
      
      return {
        success: true,
        details: `Command \`${command}\` executed successfully!\n\nOutput:\n${output || 'No output'}`
      };
    } catch (error) {
      console.log(`Verification failed with error: ${error.message}`);
      console.log(`Stderr: ${error.stderr?.toString() || 'No stderr'}`);
      console.log(`Stdout: ${error.stdout?.toString() || 'No stdout'}`);
      
      // Parse errors from the output
      const errorOutput = error.stderr?.toString() || error.stdout?.toString() || '';
      const parsedErrors = parseErrorOutput(errorOutput, filePath, fileType);
      
      return {
        success: false,
        details: `Command \`${command}\` failed!\n\nError:\n${errorOutput}`,
        errors: parsedErrors
      };
    }
  } catch (error) {
    console.error(`Error in verifyFix: ${error}`);
    return {
      success: false,
      details: `Exception during verification: ${error}`,
      errors: []
    };
  }
}

// Helper function to parse error output into a structured format
function parseErrorOutput(output, filePath, fileType) {
  try {
    const errors = [];
    const lines = output.split('\n');
    
    if (fileType === 'JavaScript/TypeScript') {
      // Parse ESLint-like errors
      for (const line of lines) {
        if (line.includes(filePath) && (line.includes('error') || line.includes('warning'))) {
          errors.push(line.trim());
        }
      }
    } else if (fileType === 'Python') {
      // Parse Pylint-like errors
      for (const line of lines) {
        if (line.includes(filePath) && (line.includes('E:') || line.includes('W:'))) {
          errors.push(line.trim());
        }
      }
    } else {
      // Generic error parsing - just include lines that mention the file path
      for (const line of lines) {
        if (line.includes(filePath)) {
          errors.push(line.trim());
        }
      }
    }
    
    // If we couldn't find specific errors mentioning the file, include some context
    if (errors.length === 0 && output.trim()) {
      const relevantLines = output.split('\n')
        .filter(line => line.trim() && (line.includes('error') || line.includes('fail')))
        .slice(0, 10); // Limit to 10 error lines
      
      errors.push(...relevantLines.map(line => line.trim()));
    }
    
    return errors;
  } catch (error) {
    console.error(`Error parsing error output: ${error}`);
    return [output.trim().substring(0, 500)]; // Return truncated output as a single error
  }
}

// Run the main function
fixCode().catch(error => {
  console.error('Uncaught error in fixCode:', error);
  process.exit(1);
});

File: .github/scripts/fix_code.py
#!/usr/bin/env python3
"""
Script to fix code issues using OpenAI.
"""
import os
import json
import re
import subprocess
import requests

def get_file_content(file_path):
    """Read and return file content."""
    with open(file_path, 'r') as f:
        return f.read()

def write_file_content(file_path, content):
    """Write content to file."""
    with open(file_path, 'w') as f:
        f.write(content)

def get_file_type(file_path):
    """Determine file type based on extension."""
    ext = os.path.splitext(file_path)[1].lower()
    
    if ext in ['.py']:
        return 'Python'
    elif ext in ['.js', '.jsx', '.ts', '.tsx']:
        return 'JavaScript/TypeScript'
    elif ext in ['.java']:
        return 'Java'
    elif ext in ['.rb']:
        return 'Ruby'
    elif ext in ['.go']:
        return 'Go'
    elif ext in ['.php']:
        return 'PHP'
    elif ext in ['.c', '.cpp', '.h', '.hpp']:
        return 'C/C++'
    elif ext in ['.cs']:
        return 'C#'
    elif ext in ['.html', '.htm']:
        return 'HTML'
    elif ext in ['.css', '.scss', '.sass', '.less']:
        return 'CSS'
    else:
        return 'Unknown'

def get_errors_for_file(file_path):
    """Extract errors for a specific file from environment variables."""
    try:
        # Run pylint and pytest to get errors for this file
        lint_errors = run_lint_for_file(file_path)
        test_errors = run_tests_for_file(file_path)
        
        all_errors = lint_errors + test_errors
        return all_errors
    except Exception as e:
        print(f"Error getting errors for {file_path}: {e}")
        return []

def run_lint_for_file(file_path):
    """Run linting for a specific file and return errors."""
    errors = []
    try:
        result = subprocess.run(
            ["pylint", file_path, "--exit-zero", "--output-format=text"], 
            capture_output=True, 
            text=True,
            check=False
        )
        
        # Parse pylint output
        lint_errors = re.findall(r'([^:]+):(\d+):(\d+): ([A-Z]\d+): (.+)', result.stdout)
        for match in lint_errors:
            _, line, col, code, message = match
            errors.append(f"Lint error {code} at line {line}: {message}")
    except Exception as e:
        print(f"Error running pylint for {file_path}: {e}")
    
    return errors

def run_tests_for_file(file_path):
    """Run tests that might be affected by this file and return errors."""
    errors = []
    try:
        # Determine if this is a module file or test file
        is_test_file = 'test_' in os.path.basename(file_path) or '/tests/' in file_path
        
        if is_test_file:
            # If it's a test file, just run that test
            test_path = file_path
        else:
            # If it's a module file, try to find corresponding test files
            module_name = os.path.splitext(os.path.basename(file_path))[0]
            test_path = f"tests/test_{module_name}.py"
            if not os.path.exists(test_path):
                # If no direct test file, run all tests
                test_path = "tests/"
        
        # Run pytest on the identified test path
        result = subprocess.run(
            ["python", "-m", "pytest", test_path, "-v"], 
            capture_output=True, 
            text=True,
            check=False
        )
        
        # Parse pytest output for errors
        if "FAILURES" in result.stdout:
            # Extract failed test information
            failed_tests = re.findall(r"(FAILED\s+[^\n]+)", result.stdout)
            error_messages = re.findall(r"E\s+([^\n]+)", result.stdout)
            
            # Match errors to this file
            module_name = os.path.splitext(os.path.basename(file_path))[0]
            
            # Add relevant test failures
            for test in failed_tests:
                if module_name in test or is_test_file:
                    errors.append(f"Test failure: {test}")
            
            # Add error messages for context
            if errors and error_messages:
                errors.extend([f"Error message: {msg}" for msg in error_messages[:5]])  # Limit to first 5 messages
    except Exception as e:
        print(f"Error running tests for {file_path}: {e}")
    
    return errors

def fix_file_with_openai(file_path, file_content, errors, file_type, attempt):
    """Use OpenAI to fix errors in the file."""
    api_key = os.environ.get('OPENAI_API_KEY')
    if not api_key:
        raise ValueError("OpenAI API key not found in environment variables")
    
    # Construct the prompt
    error_text = "\n".join(errors)
    
    prompt = f"""You are an expert {file_type} developer. I need your help fixing errors in a file.

File Path: {file_path}
File Type: {file_type}
Attempt: {attempt}

The file has the following errors:
{error_text}

Here is the current content of the file:
```
{file_content}
```

Please provide ONLY the fixed version of the file with no explanation. Your response should be the complete file content that resolves the errors."""

    # Make the API request
    response = requests.post(
        'https://api.openai.com/v1/chat/completions',
        headers={
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {api_key}'
        },
        json={
            'model': 'gpt-3.5-turbo',
            'messages': [{'role': 'user', 'content': prompt}],
            'temperature': 0.2,
            'max_tokens': 4096
        }
    )
    
    if response.status_code != 200:
        raise ValueError(f"OpenAI API request failed: {response.text}")
    
    # Extract the generated code
    result = response.json()
    if not result.get('choices') or not result['choices'][0].get('message'):
        raise ValueError("Unexpected response format from OpenAI")
    
    fixed_content = result['choices'][0]['message']['content']
    
    # Strip out markdown code blocks if present
    if '```' in fixed_content:
        match = re.search(r'```(?:\w+)?\n([\s\S]+?)\n```', fixed_content)
        if match:
            fixed_content = match.group(1)
    
    return fixed_content

def main():
    """Main function to fix code issues."""
    # Get affected files from environment
    affected_files_json = os.environ.get('ERROR_FILES')
    max_attempts = int(os.environ.get('MAX_ATTEMPTS', '3'))
    
    if affected_files_json:
        affected_files = json.loads(affected_files_json)
    else:
        # Fallback to checking Python files that most likely have issues
        affected_files = ["calculator/calculator.py", "tests/test_calculator.py"]
    
    changes_summary = []
    
    for file_path in affected_files:
        if not os.path.exists(file_path):
            print(f"File not found: {file_path}")
            continue
        
        print(f"Attempting to fix {file_path}...")
        original_content = get_file_content(file_path)
        file_type = get_file_type(file_path)
        
        # Get errors for this file
        errors = get_errors_for_file(file_path)
        if not errors:
            print(f"No errors found for {file_path}, skipping.")
            continue
        
        print(f"Found {len(errors)} errors in {file_path}")
        
        # Try to fix the file
        fixed = False
        current_content = original_content
        
        for attempt in range(1, max_attempts + 1):
            print(f"Fix attempt {attempt}/{max_attempts} for {file_path}")
            
            try:
                # Get fix from OpenAI
                fixed_content = fix_file_with_openai(
                    file_path, current_content, errors, file_type, attempt
                )
                
                if not fixed_content or fixed_content == current_content:
                    print("No changes suggested or same content returned")
                    continue
                
                # Write fixed content
                write_file_content(file_path, fixed_content)
                print(f"Updated {file_path} with suggested fixes")
                
                # Create unified diff for logging
                diff_process = subprocess.run(
                    ["git", "diff", "--unified=3", "--", file_path],
                    capture_output=True,
                    text=True,
                    check=True
                )
                diff = diff_process.stdout
                
                # Add to changes summary
                changes_summary.append(f"## Changes to {file_path} (Attempt {attempt})\n```diff\n{diff}\n```")
                
                # Commit the changes
                subprocess.run(["git", "add", file_path], check=True)
                subprocess.run(["git", "commit", "-m", f"fix: Auto-fix attempt {attempt} for {file_path}"], check=True)
                
                # Check if fixes worked
                new_errors = get_errors_for_file(file_path)
                if not new_errors:
                    print(f"Successfully fixed all errors in {file_path}!")
                    fixed = True
                    break
                else:
                    print(f"Still {len(new_errors)} errors after fix attempt {attempt}")
                    errors = new_errors
                    current_content = fixed_content
            except Exception as e:
                print(f"Error in fix attempt {attempt} for {file_path}: {e}")
        
        if not fixed:
            print(f"Could not fix all errors in {file_path} after {max_attempts} attempts")
    
    # Set GitHub Actions output
    with open(os.environ.get("GITHUB_OUTPUT", "output.txt"), "a") as f:
        changes_output = "\n\n".join(changes_summary)
        f.write(f"changes_summary<<EOF\n{changes_output}\nEOF\n")
    
    print("Completed fix attempts.")

if __name__ == "__main__":
    main()